src/test/java/htsjdk/samtools/cram/structure/CramCompressionRecordTest.java

```diff
@@ -4,85 +4,189 @@
+
+        int expectedSpan = r.readLength;
+        assertSpanAndEnd(r, usePositionDeltaEncoding, expectedSpan);
 
         r = new CramCompressionRecord();
-        r.alignmentStart = 1;
+        r.alignmentStart = 10;
         r.readLength = readLength;
         r.setSegmentUnmapped(false);
         r.readFeatures = new ArrayList<>();
         String softClip = "AAA";
         r.readFeatures.add(new SoftClip(1, softClip.getBytes()));
-        Assert.assertEquals(r.getAlignmentEnd(), r.readLength + r.alignmentStart - 1 - softClip.length());
+
+        expectedSpan = r.readLength - softClip.length();
+        assertSpanAndEnd(r, usePositionDeltaEncoding, expectedSpan);
 
         r = new CramCompressionRecord();
-        r.alignmentStart = 1;
+        r.alignmentStart = 20;
         r.readLength = readLength;
         r.setSegmentUnmapped(false);
         r.readFeatures = new ArrayList<>();
         int deletionLength = 5;
         r.readFeatures.add(new Deletion(1, deletionLength));
-        Assert.assertEquals(r.getAlignmentEnd(), r.readLength + r.alignmentStart - 1 + deletionLength);
+
+        expectedSpan = r.readLength + deletionLength;
+        assertSpanAndEnd(r, usePositionDeltaEncoding, expectedSpan);
 
         r = new CramCompressionRecord();
-        r.alignmentStart = 1;
+        r.alignmentStart = 30;
         r.readLength = readLength;
         r.setSegmentUnmapped(false);
         r.readFeatures = new ArrayList<>();
         String insertion = "CCCCCCCCCC";
         r.readFeatures.add(new Insertion(1, insertion.getBytes()));
-        Assert.assertEquals(r.getAlignmentEnd(), r.readLength + r.alignmentStart - 1 - insertion.length());
 
+        expectedSpan = r.readLength - insertion.length();
+        assertSpanAndEnd(r, usePositionDeltaEncoding, expectedSpan);
 
         r = new CramCompressionRecord();
-        r.alignmentStart = 1;
+        r.alignmentStart = 40;
         r.readLength = readLength;
         r.setSegmentUnmapped(false);
         r.readFeatures = new ArrayList<>();
         r.readFeatures.add(new InsertBase(1, (byte) 'A'));
-        Assert.assertEquals(r.getAlignmentEnd(), r.readLength + r.alignmentStart - 1 - 1);
+
+        expectedSpan = r.readLength - 1;
+        assertSpanAndEnd(r, usePositionDeltaEncoding, expectedSpan);
+    }
+
+    private void assertSpanAndEnd(final CramCompressionRecord r,
+                                  final boolean usePositionDeltaEncoding,
+                                  final int expectedSpan) {
+        final int expectedEnd = expectedSpan + r.alignmentStart - 1;
+        Assert.assertEquals(r.getAlignmentSpan(usePositionDeltaEncoding), expectedSpan);
+        Assert.assertEquals(r.getAlignmentEnd(usePositionDeltaEncoding), expectedEnd);
     }
 
-    @Test
-    public void test_isPlaced() {
+    // show that alignmentEnd and alignmentSpan are set once only and do not update
+    // to reflect the state of the record
+
+    // TODO: is this the behavior we actually want?
```

Comment:
Ohh - I didn't see this before. Not updating the span when either start or length is updated seems dangerous and totally wrong to me. Not sure what you think, but because they're publicly accessible, we have to assume they're mutable, so I think getAlignmentEnd/Span needs to always calculate on demand.

This is a great example of why I feel like we need to bite the bullet and refactor the underlying classes (`CramCompressionRecord` in this case) in order to make progress and write reasonable tests with confidence, rather than making incremental changes and point fixes. Its really hard to know what to do in cases like this.

Reply:
In looking back at previous versions, it looks like this behavior always existed and isn't new, so we can defer a fix to another PR.